File: msad2022_pri/prototyping/testTraceCamTilt.py
import sys
sys.path.append("/usr/local/lib/python3.9/site-packages")
import cv2
import math
import numpy as np
import time

# constants
FRAME_WIDTH = 640
FRAME_HEIGHT = 480

# callback function for trackbars
def nothing(x):
    pass

# prepare the camera
cap = cv2.VideoCapture(0)
cap.set(cv2.CAP_PROP_FRAME_WIDTH,FRAME_WIDTH)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT,FRAME_HEIGHT)
cap.set(cv2.CAP_PROP_FPS,90)

# create trackbars
cv2.namedWindow("testTrace1")

cv2.createTrackbar("H_min", "testTrace1", 0, 179, nothing)       # max Hue is 179
cv2.createTrackbar("H_max", "testTrace1", 179, 179, nothing)
cv2.createTrackbar("S_min", "testTrace1", 0, 255, nothing)
cv2.createTrackbar("S_max", "testTrace1", 255, 255, nothing)
cv2.createTrackbar("V_min", "testTrace1", 130, 255, nothing)
cv2.createTrackbar("V_max", "testTrace1", 255, 255, nothing)
cv2.createTrackbar("Edge",  "testTrace1", 0, 1, nothing)

while True:
    # obtain values from the trackbars
    h_min = cv2.getTrackbarPos("H_min", "testTrace1")
    h_max = cv2.getTrackbarPos("H_max", "testTrace1")
    s_min = cv2.getTrackbarPos("S_min", "testTrace1")
    s_max = cv2.getTrackbarPos("S_max", "testTrace1")
    v_min = cv2.getTrackbarPos("V_min", "testTrace1")
    v_max = cv2.getTrackbarPos("V_max", "testTrace1")
    edge  = cv2.getTrackbarPos("Edge",  "testTrace1")

    dx = int(FRAME_WIDTH/2)

    time.sleep(0.01)

    ret, img = cap.read()
    # crop the nearest/bottom part of the image
    img_bgr = img[int(FRAME_HEIGHT/2):FRAME_HEIGHT, 0:FRAME_WIDTH]
    # reduce the noise
    img_med = cv2.medianBlur(img_bgr, 5)
    # convert the image from BGR to HSV
    img_hsv = cv2.cvtColor(img_med, cv2.COLOR_BGR2HSV)
    # binarize the image
    img_bin = cv2.inRange(img_hsv, (h_min, s_min, v_min), (h_max, s_max, v_max))
    # inverse the image
    img_inv = cv2.bitwise_not(img_bin)
    # mask the image
    img_tgt = cv2.bitwise_and(img_hsv, img_hsv, mask=img_inv)
    # label connected components in the image
    num_labels, img_labeled, stats, centroids = cv2.connectedComponentsWithStats(img_inv)

    # ignore the largest black part (the first label)
    stats = np.delete(stats, 0, 0)
    centroids = np.delete(centroids, 0, 0)
    num_labels = num_labels - 1

    # prepare empty matrices
    img_lbl = np.zeros_like(img_bin)
    img_edge_cvt = np.zeros_like(img_bgr)
    
    if num_labels >= 1:
        # obtain the index to the label which has the largest area
        index = np.argmax(stats[:,4])
        # obtain the attributes of the label
        x = stats[index][0]
        y = stats[index][1]
        w = stats[index][2]
        h = stats[index][3]
        s = stats[index][4]
        mx = int(centroids[index][0])
        my = int(centroids[index][1])
        # draw a circle at the centroid on the target image
        cv2.circle(img_tgt, (mx, my), 20, (179,0,255), -1)
        # copy the largest component in the empty matrix
        img_lbl[img_labeled == index+1, ] = 255
        # detect edges
        img_edge = cv2.Canny(img_lbl, 255, 255)
        # convert the binary image into a regular RGB image
        img_edge_cvt = np.where(img_edge[..., np.newaxis] == 255, (255,255,255), (0,0,0)).astype(np.uint8)
        # find contours
        contours, hierarchy = cv2.findContours(img_edge, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        if len(contours) >= 2:
            # derive an approximated line for either edge by regression
            vx,vy,x0,y0 = cv2.fitLine(contours[edge], cv2.DIST_L2, 0, 0.01, 0.01)
            m = FRAME_WIDTH
            x1 = int(x0 - m*vx)
            y1 = int(y0 - m*vy)
            x2 = int(x0 + m*vx)
            y2 = int(y0 + m*vy)
            retval, p1, p2 = cv2.clipLine((0,0,img_tgt.shape[1],img_tgt.shape[0]), (x1,y1), (x2,y2))
            # draw the line
            #print(mx, p1, p2)
            cv2.line(img_tgt, p1, p2, (179,0,255), 10)
            # adjust the centroid by the line tilt
            if vy != 0:
                mx = mx - p1[0] + p2[0]
            # indicate the adjusted centroid by the vertical line
            cv2.line(img_tgt, (mx, 0), (mx, int(FRAME_HEIGHT/2)), (179,0,255), 10)
            dx = int(FRAME_WIDTH/2) - mx
            #print(f"dx = {dx}")

    # concatinate the images - original + edge + target
    img_v = cv2.vconcat([img,img_edge_cvt,img_tgt])
    # shrink the concatinated image to avoid delay in transmission
    img_comm = cv2.resize(img_v, (int(img_v.shape[1]/4), int(img_v.shape[0]/4)))
    # transmit and display the image
    cv2.imshow("testTrace2", img_comm)

    c = cv2.waitKey(1) # show the window
    if c == ord('q') or c == ord('Q'):
        break

cv2.destroyAllWindows

File: msad2022_pri/prototyping/testTraceCamTilt.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <thread>

using namespace std;
using namespace cv;
using std::this_thread::sleep_for;

#define FRAME_WIDTH  640
#define FRAME_HEIGHT 480

int hmin=0,hmax=179,smin=0,smax=255,vmin=130,vmax=255,edge=0;

int main() {
  /* prepare the camera */
  VideoCapture cap(0);
  cap.set(CAP_PROP_FRAME_WIDTH,FRAME_WIDTH);
  cap.set(CAP_PROP_FRAME_HEIGHT,FRAME_HEIGHT);
  cap.set(CAP_PROP_FPS,90);

  if (!cap.isOpened()) {
    cout << "cap is not open" << endl;
  }

  /* create trackbars */
  namedWindow("testTrace1");
  createTrackbar("H_min", "testTrace1", nullptr, 179, nullptr);
  setTrackbarPos("H_min", "testTrace1", hmin);
  createTrackbar("H_max", "testTrace1", nullptr, 179, nullptr);
  setTrackbarPos("H_max", "testTrace1", hmax);
  createTrackbar("S_min", "testTrace1", nullptr, 255, nullptr);
  setTrackbarPos("S_min", "testTrace1", smin);
  createTrackbar("S_max", "testTrace1", nullptr, 255, nullptr);
  setTrackbarPos("S_max", "testTrace1", smax);
  createTrackbar("V_min", "testTrace1", nullptr, 255, nullptr);
  setTrackbarPos("V_min", "testTrace1", vmin);
  createTrackbar("V_max", "testTrace1", nullptr, 255, nullptr);
  setTrackbarPos("V_max", "testTrace1", vmax);
  createTrackbar("Edge",  "testTrace1", nullptr, 1, nullptr);
  setTrackbarPos("Edge",  "testTrace1", edge);

  while (true) {
    /* obtain values from the trackbars */
    hmin = getTrackbarPos("H_min", "testTrace1");
    hmax = getTrackbarPos("H_max", "testTrace1");
    smin = getTrackbarPos("S_min", "testTrace1");
    smax = getTrackbarPos("S_max", "testTrace1");
    vmin = getTrackbarPos("V_min", "testTrace1");
    vmax = getTrackbarPos("V_max", "testTrace1");
    edge = getTrackbarPos("Edge",  "testTrace1");

    Mat img, img_med, img_hsv, img_bin, img_inv, img_tgt, img_edge, img_comm;
    int dx = FRAME_WIDTH/2, c;

    sleep_for(chrono::milliseconds(10));

    cap.read(img);
    /* crop the nearest/bottom part of the image */
    Mat img_bgr(img, Rect(0,FRAME_HEIGHT/2,FRAME_WIDTH,FRAME_HEIGHT/2));
    /* reduce the noise */
    medianBlur(img_bgr, img_med, 5);
    /* convert the image from BGR to HSV */
    cvtColor(img_med, img_hsv, COLOR_BGR2HSV);
    /* binarize the image */
    inRange(img_hsv, Scalar(hmin,smin,vmin), Scalar(hmax,smax,vmax), img_bin);
    /* inverse the image */
    bitwise_not(img_bin, img_inv);
    /* mask the image */
    bitwise_and(img_hsv, img_hsv, img_tgt, img_inv);
    /* label connected components in the image */
    Mat img_labeled, stats, centroids;
    int num_labels = connectedComponentsWithStats(img_inv, img_labeled, stats, centroids);

    /* ignore the largest black part (the first label) */
    Mat stats_rest, centroids_rest;
    for (int i = 1; i < num_labels; i++) {
      stats_rest.push_back(stats.row(i));
      centroids_rest.push_back(centroids.row(i));
    }
    num_labels--;

    /* prepare empty matrices */
    Mat img_lbl = Mat::zeros(img_bin.size(), CV_8UC1);
    Mat img_edge_cvt = Mat::zeros(img_bin.size(), CV_8UC3);
    
    if (num_labels >= 1) {
      /* obtain the index to the label which has the largest area,
	 corresponding to
	 index = np.argmax(stats[:,4]) in Python */
      int index, area_largest = 0;
      for (int i = 0; i < num_labels; i++) {
        if (stats_rest.at<int>(i,4) > area_largest) {
          index = i;
          area_largest = stats_rest.at<int>(i,4);
        }
      }
      /* obtain the attributes of the label */
      int x = stats_rest.at<int>(index,0);
      int y = stats_rest.at<int>(index,1);
      int w = stats_rest.at<int>(index,2);
      int h = stats_rest.at<int>(index,3);
      int s = stats_rest.at<int>(index,4);
      int mx = centroids_rest.at<double>(index,0);
      int my = centroids_rest.at<double>(index,1);
      /* draw a circle at the centroid on the target image */
      circle(img_tgt, Point(mx, my), 20, Scalar(179,0,255), -1);
      /* copy the largest component in the empty matrix */
      for (int i = 0; i < img_labeled.size().height; i++) {
	for (int j = 0; j < img_labeled.size().width; j++) {
	  if (img_labeled.at<int>(i,j) == index+1) { /* type = CV_32S */
	    img_lbl.at<uchar>(i,j) = 255; /* type = CV_8U */
	  }
	}
      }
      /* detect edges */
      Canny(img_lbl, img_edge, 255, 255);
      /* convert the inverted binary image into a regular RGB image
	 from CV_8U to CV_8UC3 */
      int from_to[] = {0,0, 0,1, 0,2};
      mixChannels(&img_edge, 1, &img_edge_cvt, 1, from_to, 3);
      /* find contours */
      vector<vector<Point>> contours;
      vector<Vec4i> hierarchy;
      findContours(img_edge, contours, hierarchy, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);
      if (contours.size() >= 2) {
	/* derive an approximated line for either edge by regression */
	Vec4f lin; /* vx, vy, x0, y0 */
	fitLine(contours[edge], lin, DIST_L2, 0, 0.01, 0.01);
	float m = FRAME_WIDTH;
	Point p1,p2;
	p1.x = (int)(lin[2] - m*lin[0]);
	p1.y = (int)(lin[3] - m*lin[1]);
	p2.x = (int)(lin[2] + m*lin[0]);
	p2.y = (int)(lin[3] + m*lin[1]);
	clipLine(Rect(0,0,img_tgt.size().width,img_tgt.size().height), p1, p2);
	/* draw the line on the target image */
	line(img_tgt, p1, p2, Scalar(179,0,255), 10, LINE_4);
	/* adjust the centroid by the line tilt */
	if (lin[1] != 0) {
	  mx = mx - p1.x + p2.x;
	}
	/* indicate the adjusted centroid by the vertical line */
	line(img_tgt, Point(mx,0), Point(mx,FRAME_HEIGHT/2), Scalar(179,0,255), 10, LINE_4);
	dx = (FRAME_WIDTH/2) - mx;
	cout << "dx = " << dx << endl;
      }
    }

    /* concatinate the images - original + edge + target */
    Mat img_v, img_vv;
    vconcat(img, img_edge_cvt, img_v);
    vconcat(img_v, img_tgt, img_vv);
    /* shrink the concatinated image to avoid delay in transmission */
    resize(img_vv, img_comm, Size(), 0.25, 0.25);
    /* transmit and display the image */
    imshow("testTrace2", img_comm);

    c = waitKey(1);
    if ( c == 'q' || c == 'Q' ) break;
  }

  destroyAllWindows();
  return 0;
}
